### PR TITLE
support to broadcast inputs

### PR DESCRIPTION
**Pull request by zhangxiang1993**

### CODE REVIEW

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 1]
```diff
@@ -27,14 +28,18 @@ class SoftmaxXentWithLogitsShapeHelper : public ShapeHelper {
...
       const InitializationHelper* initialization_helper) const override {
-    const Tensor& logits = ctx->input(0);
-    const Tensor& labels = ctx->input(1);
-    // logits must have the same shape as labels
-    CHECK(logits.shape() == labels.shape());
+    const Tensor& logits_in = ctx->input(0);
+    const Tensor& labels_in = ctx->input(1);
 
-    TensorShape outputShape({logits.dim_size(0), 1});
+    BCast bcast(BCast::FromShape(logits_in.shape()),
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: You have the same logic in the init helper. Since you need to broadcast in the init helper anyway, the output shapes should probably be computed in the init helper, and `GetOutputShapes()` should only be a thin wrapper than fetches them and returns them.

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 2]
```diff
@@ -111,23 +163,23 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
     auto sum_exp = dml::Reduce(exp_shifted_logits, DML_REDUCE_FUNCTION_SUM,
                                reduce_strides);
 
-    // log(sum(exp(logits - max_logits)))
+    //// log(sum(exp(logits - max_logits)))
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: nit: 4 slashes

```suggestion
    // log(sum(exp(logits - max_logits)))
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 3]
```diff
@@ -95,14 +118,43 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
+        logits_shape.dim_size(1) == shape_in.dim_size(1) ? 0 : 1;
+    dml::Expression logits_bcast;
+    if (toBCastBatchDim || toBCastClassDim) {
+      logits_bcast = dml::Reinterpret(
+          logits, input_sizes,
+          dml::TensorDesc::Dimensions(
+              {0, 0, 1 - toBCastBatchDim, 1 - toBCastClassDim}));
+    } else {
+      logits_bcast = dml::Reinterpret(logits, input_sizes, {});
+    }
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc:134]: The dml::Reintrpret in the `else` clause is basically a no-op. You can simplify your logic by simply removing the `logits_bcast` variable and overwriting `logits` instead. This allows you to remove the `else` clause and gets rid of the uninitialized `logits_bcast` variable.

```suggestion
    if (toBCastBatchDim || toBCastClassDim) {
      logits = dml::Reinterpret(
          logits, input_sizes,
          dml::TensorDesc::Dimensions(
              {0, 0, 1 - toBCastBatchDim, 1 - toBCastClassDim}));
    }
```

Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: The dml::Reintrpret in the `else` clause is basically a no-op. You can simplify your logic by simply removing the `logits_bcast` variable and overwriting `logits` instead. This allows you to remove the `else` clause and gets rid of the uninitialized `logits_bcast` variable.

```suggestion
    if (toBCastBatchDim || toBCastClassDim) {
      logits = dml::Reinterpret(
          logits, input_sizes,
          dml::TensorDesc::Dimensions(
              {0, 0, 1 - toBCastBatchDim, 1 - toBCastClassDim}));
    }
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 4]
```diff
@@ -95,14 +118,43 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
+    toBCastClassDim = labels_shape.dim_size(1) == shape_in.dim_size(1) ? 0 : 1;
+
+    if (toBCastBatchDim || toBCastClassDim) {
+      labels_bcast = dml::Reinterpret(
+          labels, input_sizes,
+          dml::TensorDesc::Dimensions(
+              {0, 0, 1 - toBCastBatchDim, 1 - toBCastClassDim}));
+    } else {
+      labels_bcast = dml::Reinterpret(labels, input_sizes, {});
+    }
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]: Same here.

```suggestion
    toBCastBatchDim = labels_shape.dim_size(0) == shape_in.dim_size(0) ? 0 : 1;
    toBCastClassDim = labels_shape.dim_size(1) == shape_in.dim_size(1) ? 0 : 1;

    if (toBCastBatchDim || toBCastClassDim) {
      labels = dml::Reinterpret(
          labels, input_sizes,
          dml::TensorDesc::Dimensions(
              {0, 0, 1 - toBCastBatchDim, 1 - toBCastClassDim}));
    }
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 5]
```diff
@@ -95,14 +118,43 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
+
     // max_logits along classes.
     auto logits_max =
-        dml::Reduce(logits, DML_REDUCE_FUNCTION_MAX, reduce_strides);
+        dml::Reduce(logits_bcast, DML_REDUCE_FUNCTION_MAX, reduce_strides);
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]:
```suggestion
        dml::Reduce(logits, DML_REDUCE_FUNCTION_MAX, reduce_strides);
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 6]
```diff
@@ -95,14 +118,43 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
     auto logits_max_bcast =
         dml::Reinterpret(logits_max, input_sizes, broadcast_c_strides);
 
     // logits - max_logits.
-    auto shifted_logits = logits - logits_max_bcast;
+    auto shifted_logits = logits_bcast - logits_max_bcast;
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]:
```suggestion
    auto shifted_logits = logits - logits_max_bcast;
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 7]
```diff
@@ -111,23 +163,23 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
+    //// log(sum(exp(logits - max_logits)))
     auto log_sum_exp = dml::Log(sum_exp);
     auto log_sum_exp_bcast =
         dml::Reinterpret(log_sum_exp, input_sizes, broadcast_c_strides);
 
     // sum(-labels *
     //    ((logits - max_logits) - log(sum(exp(logits - max_logits)))))
     auto sub = log_sum_exp_bcast - shifted_logits;
-    auto mul = labels * sub;
+    auto mul = labels_bcast * sub;
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]:
```suggestion
    auto mul = labels * sub;
``` #Resolved

File: tensorflow/core/kernels/dml_softmax_ops.cc [hunk 8]
```diff
@@ -111,23 +163,23 @@ class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
...
     auto loss = dml::Reduce(mul, DML_REDUCE_FUNCTION_SUM, reduce_strides);
 
     // backprop: prob - labels, where
     //   prob = exp(logits - max_logits) / sum(exp(logits - max_logits))
     //     (where the division broadcasts along the batch dimension)
     auto sum_exp_bcast =
         dml::Reinterpret(sum_exp, input_sizes, broadcast_c_strides);
-    auto backprop = exp_shifted_logits / sum_exp_bcast - labels;
+    auto backprop = exp_shifted_logits / sum_exp_bcast - labels_bcast;
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/dml_softmax_ops.cc]:
```suggestion
    auto backprop = exp_shifted_logits / sum_exp_bcast - labels;
``` #Resolved

**Review by PatriceVignola — APPROVED** [review]
:shipit:
